designs/spot-consolidation.md

```diff
@@ -0,0 +1,105 @@
+
+This approach is predicated on customers configuring instance family flexibility, and relatively similar spot prices within a given size. For a customer flexible to all instance families within c,m,r categories, PCO would have a broad range of instance types to choose from, and would be unlikely to be forced to choose an unhealthy pool. This assumption follows EC2 spot best practices, and is relied upon by customers who use attribute-based selection with price protection.
+
+For customers who rely on instance size flexibility within a single family, or lack sufficient flexibility, they may face increased interruption rates compared to options 1 and 2. Due to the natural price differences between different instance sizes within the same family, this approach would limit flexibility to one or two instances sizes, depending on current spot markets. However, having limited flexibility goes against spot best practices, and the customer should expect increased interruptions.
+
+See appendix for examples.
+
+* 👍👍 Limits consolidation from regressing to a lowest price allocation strategy
+* 👍 Extensible to new methods of capacity allocation, since Karpenter core is now completely agnostic to spot
+* 👍 Prevents consolidation from interrupting workloads for marginal improvement
+* 👎 During a spot capacity crunch, if many capacity pools within a price range are unhealthy, the customer may get interrupted soon after the consolidation. This is exacerbated if the customer cannot tolerate instance type diversity. 
+* 👎 Tuning this value to migitage both marginal-improvement-churn and lowest-price-regression is challenging, and desired values may conflict.
+
+Note: Regardless of the decision made to solve the spot consolidation problem, we’d likely want to implement a price improvement in the future to prevent consolidation from interrupting nodes to make marginal improvements.
+
+### 3. Optimistically Launch Replacements
+
+Karpenter could optimistically launch new EC2 instances to probe PCO decision making. If PCO launches a cheaper instance type from a CreateFleet request that include the current instance type, we know that the instance is both lower price and lower interruption rate. If the PCO algorithm selects the current instance type of the running instance, Karpenter would terminate the new instance and leave the current instance running. This approach would need to include an optimistic launch cache to minimize the number of optimistic launches.
+
+* 👍👍👍 Prevents consolidation from regressing to a lowest price allocation strategy
+* 👎 May not make sense for all cloud providers
+* 👎 May consolidate to make marginal improvements, resulting in unnecessary disruption
+* 👎 Optimistic launches may confuse customers looking deeply at their bill or audit logs
+* 👎 Bugs with this mechanism are higher risk, since real instances are being launched.
```

Comment:
Should we mention the cost component of this?

Reply:
I'm not sure cost is a major factor, unless there's a bug that causes them to not terminate for some reason.